app/src/main/java/com/example/caloriecounter/feature/restaurant/ui/categories/CategoryListScreen.kt

```diff
@@ -0,0 +1,86 @@
+import androidx.compose.foundation.layout.*
+import androidx.compose.foundation.lazy.LazyColumn
+import androidx.compose.foundation.lazy.items
+import androidx.compose.material.CircularProgressIndicator
+import androidx.compose.material.MaterialTheme
+import androidx.compose.material.Text
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.LaunchedEffect
+import androidx.compose.ui.Alignment
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.res.stringResource
+import androidx.compose.ui.text.style.TextAlign
+import androidx.compose.ui.unit.dp
+import kotlinx.coroutines.flow.launchIn
+import kotlinx.coroutines.flow.onEach
+import com.example.caloriecounter.R
+import com.example.caloriecounter.base.DataState
+import com.example.caloriecounter.ui.ErrorScreen
+import com.example.domain.models.Category
+import kotlinx.coroutines.flow.Flow
+
+@Composable
+fun CategoryListScreen(
+    categories: DataState<List<Category>>,
+    effectFlow: Flow<CategoriesContract.Effect>,
+    onEventSent: (event: CategoriesContract.Event) -> Unit,
+    onNavigationRequested: (category: String) -> Unit,
+) {
+    LaunchedEffect(Unit) {
+        effectFlow.onEach { effect ->
+            when (effect) {
+                is CategoriesContract.Effect.Navigation.ToCategoryMeals -> {
+                    onNavigationRequested(effect.category)
+                }
+            }
+        }.launchIn(this)
+    }
+
+    Box(modifier = Modifier.fillMaxSize()) {
+        LazyColumn(modifier = Modifier.fillMaxSize()) {
+            item {
+                Text(
+                    text = stringResource(id = R.string.categories),
+                    modifier = Modifier.fillMaxWidth(),
+                    style = MaterialTheme.typography.h5,
+                    textAlign = TextAlign.Center,
+                    color = MaterialTheme.colors.primary
+                )
+            }
+
+            items(items = categories.data, key = { it.id }) { category ->
+                CategoryItemComponent(
+                    category = category,
+                    onClick = { onEventSent(CategoriesContract.Event.SelectedCategory(it)) },
+                    modifier = Modifier.padding(vertical = 8.dp, horizontal = 16.dp)
+                )
+            }
+        }
+
+        AnimatedVisibility(
+            visible = !categories.errorMessage.isNullOrEmpty(),
+            modifier = Modifier.align(Alignment.Center),
+            enter = fadeIn(),
+            exit = fadeOut()
+        ) {
+            ErrorScreen(
+                errorMessage = categories.errorMessage ?: stringResource(id = R.string.error),
+                reloadData = { onEventSent(CategoriesContract.Event.ReloadData) }
+            )
+        }
+
+        AnimatedVisibility(
+            visible = categories.isLoading,
+            modifier = Modifier.align(Alignment.Center),
+            enter = fadeIn(),
+            exit = fadeOut()
```

Comment:
I think it's up to you but you can extract `LazyColumn` to be clearly visible that there are 3 states Loading, Error and Data Finished Successfully

Reply:
In this case, I can separate `Success` and `Error` screen states, but with `Loading` I will lose enter/exit animation if I move it into `when`.

Reply:
At least you can add comments `//Loading //Success //Error`. That's up to you